libyara/atoms.c

```diff
@@ -1576,8 +1576,17 @@ int yr_atoms_extract_from_string(
         *atoms = NULL;
       });
 
-      yr_atoms_list_destroy(*atoms);
-      *atoms = xor_atoms;
+    yr_atoms_list_destroy(*atoms);
+    *atoms = xor_atoms;
+
+    // Recheck the atom quality, in case we have just generated some poor atoms.
+    // https://github.com/VirusTotal/yara/issues/1172
+    for (item = *atoms; item != NULL; item = item->next)
```

Comment:
I think this should be done not only for the `xor` modifier, but for all of them. So that the minimum quality is always computed from the final atoms no matter the modifiers being used.